SOP for adding new HL7 spec (#14182)
CDCgov/prime-reportstream: prime-router/docs/standard-operating-procedures/adding-conformance-profile.md

    - For example if there is a change to PID.7 you would open the `PATIENT_RESULT` class then `PATIENT` class
      then `PID` class where the edit can be made.

#### Steps to add structure:

1. Add package to `prime-router/src/main/java/fhirengine/translation/hl7/structures/` to house structure
2. Add a package named "message". Within that, add a java class named with the HL7 message type (ex ORU_R01, ORM_O01,  
   ADT_A01, etc.). This package and class are required.
3. Add a package named "group". Within that, add java classes for the HL7 groups present in the conformance profile.
4. Add a package named "segment". Within that, add java classes for the HL7 segments present in the conformance profile.
5. Add a package named "datatype". Within that, add java classes for the HL7 datatypes present in the conformance  
   profile.

- Review the java classes in the corresponding packages for other data structures to see examples of how those classes  
  should be arranged.

#### Choosing a Message Model

Once the new structure is created the ReportStream code needs to be updated to use said structure. Steps to take are:

> 1. `HL7Reader.getMessageProfile()` returns a MessageProfile object including two properties: typeID and profileID.

Reviewer: Maybe a note that this might be subject to change (i.e. adding it to the topic/receiver instead), since we know we're not going to get the right values in MSH 21

# Adding support for a new HL7v2 conformance profile in ReportStream

This document outlines the process of integrating a HL7v2 conformance profile so that a new data model can be
processed  
by ReportStream.

## Background

ReportStream uses FHIR R4 as the internal data model. Data coming into the system is converted to FHIR then data  
output can be transformed into a variety of formats dependent on the requirements on the receiver. This design  
intentionally decouples the input from the output so that the formats of the two can be different.

### Editing existing vs Creating New

> Integrating a new conformance profile involves editing existing data structures and mappings or creating new. Modifying

Reviewer 2: ```suggestion
Integrating a new conformance profile involves editing existing data structures and mappings or creating new ones. Modifying
```

> an existing structures and mappings is preferred. But that option depends on the differences present in the new   

Reviewer 2: ```suggestion
existing structures and mappings is preferred, but that option depends on the differences present in the new   
```

> # Adding a New Conformance Profile in ReportStream

Reviewer 2: ```suggestion
# Adding support for a new HL7v2 conformance profile in ReportStream
```

> # Adding a New Conformance Profile in ReportStream

Reviewer 2: Even with our pivot to erring on the side of less explicit support for conformance profiles, we still agreed there may one day be a situation where we need to add explicit support for a conformance profile, so I think this SOP is still relevant

> This document outlines the process of integrating a conformance profile so that a new data model can be sent via  

Reviewer 2: ```suggestion
This document outlines the process of integrating a HL7v2 conformance profile so that a new data model can be processed by  
```

ReportStream.

## Background

ReportStream uses FHIR R4 as the internal data model. Data coming into the system is converted to FHIR then data
output  
can be transformed into a variety of formats dependent on the requirements on the receiver. This design intentionally  
decouples the input from the output so that the formats of the two can be different.

## Data Structures

### HL7

ReportStream uses the [HAPI HL7v2](https://github.com/hapifhir/hapi-hl7v2) library to parse HL7v2 messages. That  

> library includes structures for v2.1 to v2.8.1. When HL7 conformance profiles differ from the base version specs a new  

Reviewer 2: Also for 2.5 and 2.7? Or did you mean "from" instead of "for"?

structure is needed. Steps to add a new structure:

1. Add package to `prime-router/src/main/java/fhirengine/translation/hl7/structures/`  to house structure
2. Add a package named "message". Within that, add a java class named with the HL7 message type (ex ORU_R01, ORM_O01,
   ADT_A01, etc.). This package and class are required.
3. Add a package named "group". Within that, add java classes for the HL7 groups present in the conformance profile.
4. Add a package named "segment". Within that, add java classes for the HL7 segments present in the conformance profile.
5. Add a package named "datatype". Within that, add java classes for the HL7 datatypes present in the conformance
   profile.

> - Review the java classes in the corresponding packages for other data structures to see examples of how those classes

Author: hmmm, to my knowledge they are all the same pattern. Do you know of any differences?

> - Review the java classes in the corresponding packages for other data structures to see examples of how those classes

Author: I looked again and feel pretty confident on the top level structure here, at least for ORU_R01. It might be possible that they change with other HL7 message types(like ADT) but I think we cross that bridge later

> - Review the java classes in the corresponding packages for other data structures to see examples of how those classes

Reviewer 3: Sounds good to me!

  should be arranged.

> #### Choosing a Message Model

Reviewer 2: This comment goes for the rest of the document: Is the content here still accurate given our pivot? Is the code and folder structure mentioned here going to remain the same?

Author: Left in the content for adding new and added content for updating which covers our new direction

Once the new structure is created the ReportStream code needs to be updated to use said structure. Steps to do this:

1. `HL7Reader.getMessageProfile()` returns a MessageProfile object including two properties: typeID and profileID. Both
   properties are required to determine the correct structure. First, `typeID` is derived from `MSH-9` then `profileID`
   is looked up via `oidProfileMap`. The key for that map is the OID found in `MSH-21-3`. The expected OID needs to be
   added to that map and paired with string value to be used as the profileID.
2. The structure used to parse a given HL7 message is determined in `HL7Reader.getMessageModelClasses()`. Review that
   function to ensure the combination of `typeID` and `profileID` will return the appropriate structure.

## Mapping

> When looking to add a new mapping suite the first step to review existing mapping suites with an eye for reusability.

Reviewer 2: ```suggestion
When looking to add a new mapping suite the first step is to review existing mapping suites with an eye for reusability.
```

Mappings are structured with base mappings separate from mappings that deviate from the base specification.

> You can see the separation in the mapping for v251-elr. That conformance profile is closely aligned with the profile for

Reviewer 3: Some more clarification here would be useful for the uninitiated. Is this talking about the separation in mapping between files like [catchall](https://github.com/CDCgov/prime-reportstream/blob/master/prime-router/metadata/HL7/catchall/hl7/message/ORU_R01.yml) and [v251-elr](https://github.com/CDCgov/prime-reportstream/blob/master/prime-router/metadata/HL7/v251-elr/hl7/message/ORU_R01.yml)? These are in the metadata folder which hasn't been referenced in this document so far. Mainly, a link to where we can see this separation would be helpful.

Author: I didn't want to add to many direct links to folders as I'm fairly sure they are going to change/be deleted.

v251. That presents the opportunity for reuse.

### HL7 to FHIR

1. Review existing data type mappings located in `prime-router/metadata/HL7/datatypes/` for reuse
2. Add folder to `prime-router/metadata/HL7/` with a name indicating the new conformance profile

> 3. Copy folder `fhir` from `prime-router/metadata/HL7/v251-elr` and add to your new mapping folder.

Reviewer 3: I haven't looked at the file in this folder before. Is this just a list of all the FHIR resources that we are using in a given conformance profile? (And I suppose it would remain fairly constant across profiles if so).

Author: yep exactly right

> 4. Create folder `hl7` in your new mapping folder with the following subfolders:

Reviewer 3: v251-elr also has an extension folder, do we need to mention that here?

      update as needed.
    - "datatypes" - **optional** - this folder will hold any new datatypes or datatypes that deviate from the base
      datatypes
5. Starting with message type file in your `/message` directory build the mappings.
   The [Templating Guide](https://github.com/LinuxForHealth/hl7v2-fhir-converter/blob/master/TEMPLATING.md) can be
   helpful.

### FHIR to HL7

1. Review existing datatype mappings located in `prime-router/src/main/resources/metadata/hl7_mapping`
2. Add folder to `prime-router/src/main/resources/metadata/hl7_mapping` with a name indicating the new conformance
   profile
3. Add YML file for the HL7 message type (ex ORU_R01, ORM_O01, ADT_A01, etc)
    - This file will be the starting point for mapping
4. Create a folder named "base". This will hold folders and sub-folders for the HL7 message groups. See
   the `prime-router/src/main/resources/metadata/hl7_mapping/v251-elr/` for an example
5. Create a folder "resources". This will hold yaml files from mapping directly from FHIR resources
6. Create a folder "datatypes". This folder will hold any new datatypes or datatypes that deviate from the base
   datatypes

#### Resources

> - The FHIR to HL7 mappings extensively use [FHIR Path](https://hl7.org/fhirpath/N1/) to access values in FHIR

Reviewer 3: I don't know how out of date this doc has been made by the recent file shuffle, but might be good to link here? https://github.com/CDCgov/prime-reportstream/blob/master/prime-router/docs/design/design/mapping-schemas.md

Author: another good call, also added
